Fix #8883 Own DateTimeConverter should be considered

weissreto opened the pull request:
Converter.getType() has nothing to do with the type of the value that
the converter can handle. Possible values are for type are 'date',
'time' and 'both'.

Instead, I try to use the converter and if it is not able to convert a
value I put it on a blacklist, so that the next time the Primefaces
codes is used. This is important since the standard DateTimeConverter
does not handle java.time.* classes correct. On the other hand our own
converter should be used for our own date time data types.

weissreto committed:
Fix #8883 Own DateTimeConverter should be considered

Converter.getType() has nothing to do with the type of the value that
the converter can handle. Possible values are for type are 'date',
'time' and 'both'.

Instead, I try to use the converter and if it is not able to convert a
value I put it on a blacklist, so that the next time the Primefaces
codes is used. This is important since the standard DateTimeConverter
does not handle java.time.* classes correct. On the other hand our own
converter should be used for our own date time data types.

weissreto committed:
Fix #8883 Fix code style and indentation errors

tandraschko commented:
1) i dont like a try-error approach; CC @christophs78 
2) a static cache can lead to mem leaks, a AppScoped map would be the way to go

christophs78 commented:
Independent of the concrete solution please add a integration-test for this. This is the best way to avoid future issues regarding your kind of special usecase with custom DateTime-classes. 
(use e.g. https://github.com/primefaces/primefaces/blob/master/primefaces-integration-tests/src/test/java/org/primefaces/integrationtests/datepicker/DatePicker013Test.java and https://github.com/primefaces/primefaces/blob/master/primefaces-integration-tests/src/main/webapp/datepicker/datePicker013.xhtml as template)

christophs78 commented:
> 1. i dont like a try-error approach; CC @christophs78

Are we sure `javax.faces.convert.ConverterException` is only thrown in the case of invalid combination of class and coverter? (Or may it also thrown du to other reasons? https://docs.oracle.com/javaee/7/api/javax/faces/convert/ConverterException.html sounds a bit generic. With your modification we won´t call this converter anymore for a specific class after it has once thrown `javax.faces.convert.ConverterException`.)

> 2. a static cache can lead to mem leaks, a AppScoped map would be the way to go

weissreto commented:
I will try to provide an integration-test and also my converter. This will need some time. Hope I can work on it on Friday.

melloware commented on primefaces/src/main/java/org/primefaces/util/CalendarUtils.java:
```diff
@@ -52,6 +63,8 @@ public class CalendarUtils {
 
     private static final PatternConverter PATTERN_CONVERTER = new DateTimePatternConverter();
 
+    private static final Map<Class<?>, Set<Class<? extends Converter>>> BAD_CONVERTERS = new HashMap<>();
```
INVALID_CONVERTERS

melloware commented on primefaces/src/main/java/org/primefaces/util/CalendarUtils.java:
```diff
@@ -346,6 +356,19 @@ else if (value instanceof LocalTime) {
         }
     }
 
+    private static void addBadConverter(Converter converter, Object value) {
```
addInvalidConverter

melloware commented on primefaces/src/main/java/org/primefaces/util/CalendarUtils.java:
```diff
@@ -346,6 +356,19 @@ else if (value instanceof LocalTime) {
+        converters.add(converter.getClass());
+    }
+
+    private static boolean isGoodConverter(Converter converter, Object value) {
```
addValidConverter

weissreto commented:
> > 1. i dont like a try-error approach; CC @christophs78
> 
> Are we sure `javax.faces.convert.ConverterException` is only thrown in the case of invalid combination of class and coverter? (Or may it also thrown du to other reasons? https://docs.oracle.com/javaee/7/api/javax/faces/convert/ConverterException.html sounds a bit generic. With your modification we won´t call this converter anymore for a specific class after it has once thrown `javax.faces.convert.ConverterException`.)
> 
The method getAsString from javax.faces.convert.DateTimeConverter only throws that Exception if a given Object cannot be converted to a String. Means DateFormat.format method throws an Exception. I don't see any other reason this Exception can be raised by the getAsString method of that class.

weissreto commented:
Here is our faces-config.xml with the javax.faces.DateTime converter configuration. And the implementation of our DateTimeConverter class. Also The JavaDateToIvyMappers that is used to map java.util.Date to our own Date, Time and DateTime classes.
[DateTimeConverter.zip](https://github.com/primefaces/primefaces/files/8925274/DateTimeConverter.zip)
w
Integration Test will come soon ...

weissreto committed:
Add integration test for issue #8883

Add integration test DatePicker014Test that uses an own date time class
and an own DateTimeConverter class.

The converter is bound from the data bean to the f:dateTimeConvert tag.
Normally one would instead configuring it globally in the
faces-config.xml (But I failed to do so in the integration test).

melloware commented:
Nice Integration test!   Can you address my code review comments as well renaming "bad" and "good" to "valid" and "invalid"

weissreto committed:
Fix #8883: Rename bad/good to invalid/valid and use app scoped map

Rename addBadConverter to addInvalidConverter
Rename isGoodConverter to isValidConverter
Instead of using a global Map to store the invalid converters use an app
scoped map.

melloware commented:
OK I am rerunning the integration tests now.

weissreto commented:
Thanks for your help and merging

melloware commented:
Thanks for the contribution and integration test!

tandraschko commented:
TBH i dont think this is an good change
A ConverterException says that the current value is not convertable -> e.g. a bad input on the UI and a error message will be shown the user
we now ignore the converter forever

either we introduce a own exception type or look for a nice and correct fix

user applied converters, doesnt matter if via annotation or component attr should be respected first
